{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 1,
   "metadata": {
    "collapsed": false
   },
   "outputs": [
    {
     "data": {
      "text/html": [
       "<iframe style=\"border: 0; width: 400px; height: 339px;\" src=\"https://bandcamp.com/EmbeddedPlayer/album=2664818884/size=large/bgcol=333333/linkcol=ffffff/artwork=small/track=3028022632/transparent=true/\" seamless><a href=\"http://nekkralai.bandcamp.com/album/kliud-iau\">Kliudziau by NekKralai</a></iframe>"
      ],
      "text/plain": [
       "<IPython.core.display.HTML object>"
      ]
     },
     "execution_count": 1,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "from IPython.display import HTML\n",
    "\n",
    "embed='<iframe style=\"border: 0; width: 400px; height: 339px;\" src=\"https://bandcamp.com/EmbeddedPlayer/album=2664818884/size=large/bgcol=333333/linkcol=ffffff/artwork=small/track=3028022632/transparent=true/\" seamless><a href=\"http://nekkralai.bandcamp.com/album/kliud-iau\">Kliudziau by NekKralai</a></iframe>'\n",
    "\n",
    "HTML(embed)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "metadata": {
    "collapsed": false,
    "scrolled": false
   },
   "outputs": [],
   "source": [
    "import re\n",
    "\n",
    "## path to XML\n",
    "xml_path='../XMLs/GireWA_hkySCCS_20M_1.xml'\n",
    "\n",
    "listen=False\n",
    "\n",
    "## dict to map sequence names to sequences\n",
    "strains={}\n",
    "\n",
    "strain_order=[]\n",
    "\n",
    "## iterate over every line in the XML\n",
    "for line in open(xml_path,'r'):\n",
    "    \n",
    "    ## start listening for sequence name and sequence whenever <sequence> tag was encountered\n",
    "    if '<sequence>' in line:\n",
    "        name=''\n",
    "        listen=True\n",
    "        \n",
    "    ## stop listening for sequence name and sequence whenever </sequence> tag was encountered\n",
    "    if '</sequence>' in line:\n",
    "        name=''\n",
    "        listen=False\n",
    "    \n",
    "    ## search for sequence name\n",
    "    name_cerberus = re.search('^\\t\\t\\t<taxon idref=\"([A-Za-z0-9\\_\\.]+)\"/>$',line)\n",
    "    if name_cerberus is not None and listen==True:\n",
    "        name = name_cerberus.group(1)\n",
    "        strain_order.append(name)\n",
    "        strains[name] = ''\n",
    "    \n",
    "    ## search for sequence\n",
    "    seq_cerberus = re.search('^\\t\\t\\t([A-Z\\-]+)$',line)\n",
    "    if seq_cerberus is not None and listen==True:\n",
    "        strains[name]+=seq_cerberus.group(1)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "metadata": {
    "collapsed": false,
    "scrolled": true
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Calibrating on: KJ660347p2_Gueckedou_C07_276 276\n",
      "Calibration date: 2014-12-21\n",
      "\n",
      "\n"
     ]
    }
   ],
   "source": [
    "import datetime\n",
    "\n",
    "## only isolate C07 gives dates to the rest of Baize et al isolates that match public records\n",
    "C07_date = '2014-03-20'\n",
    "\n",
    "## iterate over all strains, find C15 sequence from Baize et al\n",
    "for strain in strains.keys():\n",
    "    ## find C07 isolate\n",
    "    if 'C07' in strain:\n",
    "        print 'Calibrating on:',strain,\n",
    "        diff=int(strain.split('_')[-1])\n",
    "        print diff\n",
    "        ## identify the XML calibration date based on known isolation day of C07 isolate\n",
    "        date_1 = datetime.datetime.strptime(C07_date,\"%Y-%m-%d\")\n",
    "\n",
    "## date when tips in the XML were calibrated\n",
    "calibration_date = date_1 + datetime.timedelta(days=diff)\n",
    "\n",
    "print 'Calibration date: %s\\n\\n'%(datetime.datetime.strftime(calibration_date,\"%Y-%m-%d\"))\n",
    "\n",
    "def realDate(diff):\n",
    "    \"\"\"Adds a number of days to the calibrated time point, returns calendar date.\"\"\"\n",
    "    return datetime.datetime.strftime(calibration_date - datetime.timedelta(days=diff),\"%Y-%m-%d\")\n",
    "\n",
    "## specify output file\n",
    "output_fasta=open('EBOV_rates_project_2016.fasta','w')\n",
    "\n",
    "## report fasta\n",
    "for strain in strain_order:\n",
    "    diff=int(strain.split('_')[-1])\n",
    "    #print '>%s|%s\\n%s'%(strain,realDate(diff),strains[strain])\n",
    "    print>>output_fasta,'>%s|%s\\n%s'%(strain,realDate(diff),strains[strain])\n",
    "    \n",
    "output_fasta.close()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "metadata": {
    "collapsed": false,
    "scrolled": false
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "KP260799 ['KP260799']\n",
      "KP260800 ['KP260800']\n",
      "KP260801 ['KP260801']\n",
      "KP260802 ['KP260802']\n",
      "KM034555 ['KM034555']\n",
      "KM034556 ['KM034562']\n",
      "KM034557 ['KM233052']\n",
      "KM034558 ['KM233108']\n",
      "KM034559 ['KM034556']\n",
      "KM034560 ['KM233054']\n",
      "KM034561 ['KM034550']\n",
      "KM034562 ['KM034549']\n",
      "KM034563 ['KM034551']\n",
      "KM233035 ['KM233043']\n",
      "KM233036 ['KM233083']\n",
      "KM233037 ['KM233055', 'KM233068', 'KM233069']\n",
      "KM233038 ['KM233055', 'KM233068', 'KM233069']\n",
      "KM233039 ['KM233070']\n",
      "KM233040 ['KM034559']\n",
      "KM233041 ['KM233086']\n",
      "KM233042 ['KM233046']\n",
      "KM233043 ['KM233104']\n",
      "KM233044 ['KM233057']\n",
      "KM233045 ['KM233084']\n",
      "KM233046 ['KM233100']\n",
      "KM233047 ['KM233035']\n",
      "KM233048 ['KM233098']\n",
      "KM233049 ['KM034560']\n",
      "KM233050 ['KM233049']\n",
      "KM233051 ['KM233050']\n",
      "KM233052 ['KM233051']\n",
      "KM233053 ['KM233061']\n",
      "KM233054 ['KM233072']\n",
      "KM233055 ['KM233092']\n",
      "KM233056 ['KM233062']\n",
      "KM233057 ['KM233065', 'KM233066']\n",
      "KM233058 ['KM233109']\n",
      "KM233059 ['KM233110']\n",
      "KM233060 ['KM233111']\n",
      "KM233061 ['KM233089']\n",
      "KM233062 ['KM233038']\n",
      "KM233063 ['KM233065', 'KM233066']\n",
      "KM233064 ['KM233102']\n",
      "KM233065 ['KM233039', 'KM233040']\n",
      "KM233066 ['KM233095']\n",
      "KM233067 ['KM233047']\n",
      "KM233068 ['KM233039', 'KM233040']\n",
      "KM233069 ['KM233075', 'KM233078']\n",
      "KM233070 ['KM233075', 'KM233078']\n",
      "KM233071 ['KM034561']\n",
      "KM233072 ['KM233081']\n",
      "KM233073 ['KM233044', 'KM233045']\n",
      "KM233074 ['KM233090']\n",
      "KM233075 ['KM233048']\n",
      "KM233076 ['KM233088']\n",
      "KM233077 ['KM233093']\n",
      "KM233078 ['KM233044', 'KM233045']\n",
      "KM233079 ['KM233085']\n",
      "KM233080 ['KM233073']\n",
      "KM233081 ['KM233041', 'KM233042']\n",
      "KM233082 ['KM233114']\n",
      "KM233083 ['KM233115']\n",
      "KM233084 ['KM233036']\n",
      "KM233085 ['KM233094']\n",
      "KM233086 ['KM233105']\n",
      "KM233087 ['KM233106']\n",
      "KM233088 ['KM233037']\n",
      "KM233089 ['KM233067']\n",
      "KM233090 ['KM233118']\n",
      "KM233091 ['KM233087']\n",
      "KM233092 ['KM233082']\n",
      "KM233093 ['KM233101']\n",
      "KM233094 ['KM233041', 'KM233042']\n",
      "KM233095 ['KM233103']\n",
      "KM233096 ['KM233080']\n",
      "KM233097 ['KM233074']\n",
      "KM233098 ['KM233059']\n",
      "KM233099 ['KM233071']\n",
      "KM233100 ['KM034563']\n",
      "KM233101 ['KM233097']\n",
      "KM233102 ['KM233063']\n",
      "KM233103 ['KM233060']\n",
      "KM233104 ['KM233091']\n",
      "KM233105 ['KM233056']\n",
      "KM233106 ['KM233064']\n",
      "KM233107 ['KM233096']\n",
      "KM233108 ['KM233055', 'KM233068', 'KM233069']\n",
      "KM233109 ['KM233076']\n",
      "KM233110 ['KM233053']\n",
      "KM233111 ['KM233112']\n",
      "KM233112 ['KM233058']\n",
      "KM233113 ['KM233077']\n",
      "KM233114 ['KM233079']\n",
      "KM233115 ['KM233099']\n",
      "KM233116 ['KM034558']\n",
      "KM233117 ['KM233113']\n",
      "KM233118 ['KM233116']\n",
      "KM034549 ['KM034552']\n",
      "KM034550 ['KM034553']\n",
      "KM034551 ['KM034557']\n",
      "KM034552 ['KM233117']\n",
      "KM034553 ['KM233107']\n",
      "KM034554 ['KM034554']\n",
      "KJ660346 ['KJ660346']\n",
      "KJ660347 ['KJ660347']\n",
      "KJ660348 ['KJ660348']\n"
     ]
    }
   ],
   "source": [
    "import re\n",
    "\n",
    "xml_fasta='../alignments/EBOV_rates_project_full.fasta'\n",
    "correct_fasta='../alignments/Gire_Mali_old.fasta'\n",
    "\n",
    "xml_full_name={}\n",
    "correct_full_name={}\n",
    "\n",
    "xml_alignment={}\n",
    "correct_alignment={}\n",
    "\n",
    "xml_dates={}\n",
    "correct_dates={}\n",
    "\n",
    "xml_order=[]\n",
    "\n",
    "## import XML alignment\n",
    "for line in open(xml_fasta,'r'):\n",
    "    l=line.strip('\\n')\n",
    "    if '>' in l:\n",
    "        cerberus=re.search('[A-Z0-9]+',l)\n",
    "        if cerberus is not None:\n",
    "            name=cerberus.group()\n",
    "        xml_full_name[name]=l[1:]\n",
    "        xml_order.append(name)\n",
    "        xml_dates[name]=l.split('|')[-1]\n",
    "\n",
    "    elif name!='':\n",
    "        seq=l\n",
    "        xml_alignment[name]=seq\n",
    "        name=''\n",
    "\n",
    "## import correct alignment\n",
    "for line in open(correct_fasta,'r'):\n",
    "    l=line.strip('\\n')\n",
    "    if '>' in l:\n",
    "        name=l[1:].split('|')[0]\n",
    "        correct_full_name[name]=l[1:]\n",
    "        correct_dates[name]=l.split('|')[-1]\n",
    "\n",
    "    elif name!='':\n",
    "        seq=l\n",
    "        correct_alignment[name]=seq\n",
    "        name=''\n",
    "        \n",
    "## open output file\n",
    "mapping_out=open('strain_map.csv','w')\n",
    "\n",
    "## add header\n",
    "print>>mapping_out,'correct accession,correct AKA,matched xml accession,matched xml AKA'\n",
    "\n",
    "## iterate over each sequence\n",
    "for i in xml_order:\n",
    "    seqA=correct_alignment[i]\n",
    "    \n",
    "    best_matches_0=[]\n",
    "    best_matches_1=[]\n",
    "    \n",
    "    ## compare against every other sequence\n",
    "    for j in xml_order:\n",
    "        seqB=xml_alignment[j]\n",
    "\n",
    "        ## get sequence mismatches\n",
    "        mismatches=[0 if seqA[k]==seqB[k] else 1 for k in range(len(seqB))]\n",
    "\n",
    "        if sum(mismatches)==0:\n",
    "            best_matches_0.append(j)\n",
    "        elif sum(mismatches)==1:\n",
    "            best_matches_1.append(j)\n",
    "    \n",
    "    print i,best_matches_0\n",
    "    \n",
    "    ## output to file\n",
    "    print>>mapping_out,'%s,%s,%s,%s'%(i,correct_full_name[i],';'.join(best_matches_0),';'.join([xml_full_name[w] for w in best_matches_0]))\n",
    "    \n",
    "mapping_out.close()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "metadata": {
    "collapsed": false
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "mismatched dates: KJ660346 KJ660346p2_Kissidougou_C15_277|2014-03-19 KJ660346|Makona-Kissidougou-C15|2014-03-17 2014-03-17 2014-03-19\n",
      "mismatched dates: KJ660348 KJ660348p2_Gueckedou_C05_279|2014-03-17 KJ660348|Makona-Gueckedou-C05|2014-03-19 2014-03-19 2014-03-17\n"
     ]
    }
   ],
   "source": [
    "## iterate over strains\n",
    "for i in correct_alignment.keys():\n",
    "    ## check for mismatched dates\n",
    "    if correct_dates[i]!=xml_dates[i]:\n",
    "        print 'mismatched dates:',i,xml_full_name[i],correct_full_name[i],correct_dates[i],xml_dates[i]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 12,
   "metadata": {
    "collapsed": false,
    "scrolled": false
   },
   "outputs": [],
   "source": [
    "import re\n",
    "\n",
    "haplotypes=[]\n",
    "\n",
    "xml_fasta='../alignments/EBOV_rates_project_full.fasta'\n",
    "correct_fasta='../alignments/Gire_Mali_old.fasta'\n",
    "\n",
    "xml_full_name={}\n",
    "correct_full_name={}\n",
    "\n",
    "xml_alignment={}\n",
    "correct_alignment={}\n",
    "\n",
    "xml_dates={}\n",
    "correct_dates={}\n",
    "\n",
    "xml_order=[]\n",
    "correct_order=[]\n",
    "\n",
    "\n",
    "## Construct haplotype map\n",
    "for line in open(correct_fasta,'r'):\n",
    "    l=line.strip('\\n')\n",
    "    if '>' in l:\n",
    "        cerberus=re.search('[A-Z0-9]+',l)\n",
    "        if cerberus is not None:\n",
    "            name=cerberus.group()\n",
    "        correct_order.append(name)\n",
    "\n",
    "\n",
    "    elif name!='':\n",
    "        ## ignore last nucleotide, because XML sequences do not have it\n",
    "        seq=l[:-1]\n",
    "        name=''\n",
    "        \n",
    "        if seq not in haplotypes:\n",
    "            haplotypes.append(seq)\n",
    "        \n",
    "\n",
    "\n",
    "## import XML alignment\n",
    "for line in open(xml_fasta,'r'):\n",
    "    l=line.strip('\\n')\n",
    "    if '>' in l:\n",
    "        cerberus=re.search('[A-Z0-9]+',l)\n",
    "        if cerberus is not None:\n",
    "            name=cerberus.group()\n",
    "        xml_full_name[name]=l[1:]\n",
    "        xml_dates[name]=l.split('|')[-1]\n",
    "        xml_order.append(name)\n",
    "        \n",
    "    elif name!='':\n",
    "        seq=l\n",
    "        xml_alignment[name]=seq\n",
    "        name=''\n",
    "        \n",
    "\n",
    "## import correct alignment\n",
    "for line in open(correct_fasta,'r'):\n",
    "    l=line.strip('\\n')\n",
    "    if '>' in l:\n",
    "        name=l[1:].split('|')[0]\n",
    "        correct_full_name[name]=l[1:]\n",
    "        correct_dates[name]=l.split('|')[-1]\n",
    "\n",
    "    elif name!='':\n",
    "        ## ignore last nucleotide, because XML sequences do not have it\n",
    "        seq=l[:-1]\n",
    "        correct_alignment[name]=seq\n",
    "        name=''\n",
    "        \n",
    "        \n",
    "\n",
    "## sort haplotypes (just in case the XML sequences were sorted that way)\n",
    "#haplotypes=sorted(haplotypes)\n",
    "\n",
    "## output assigned ID to each haplotype\n",
    "hap_IDs=open('haplotype_IDs.csv','w')\n",
    "print>>hap_IDs,'ID,sequence'\n",
    "for i,hap in enumerate(haplotypes):\n",
    "    print>>hap_IDs,'%d,%s'%(i+1,hap)\n",
    "\n",
    "hap_IDs.close()\n",
    "\n",
    "## output the haplotype ID of each sequence in the order in which strains appear in the XML\n",
    "hap_map=open('haplotype_map_xml.csv','w')\n",
    "print>>hap_map,'name,collection date,hap ID'\n",
    "for strain in xml_order:\n",
    "    name = xml_full_name[strain]\n",
    "#    print name,strain\n",
    "\n",
    "    hapID=haplotypes.index(xml_alignment[strain])+1\n",
    "    \n",
    "    print>>hap_map,'%s,%s,%s,%s'%(name,strain,xml_dates[strain],hapID)\n",
    "             \n",
    "hap_map.close()\n",
    "\n",
    "## output the haplotype ID of each sequence in the order in which strains appear in the correct alignment\n",
    "hap_map=open('haplotype_map_correct.csv','w')\n",
    "print>>hap_map,'name,collection date,hap ID'\n",
    "for strain in correct_order:\n",
    "\n",
    "    name = correct_full_name[strain]\n",
    "#    print name,strain\n",
    "\n",
    "    hapID=haplotypes.index(correct_alignment[strain])+1\n",
    "    \n",
    "    print>>hap_map,'%s,%s,%s,%s'%(name,strain,correct_dates[strain],hapID)\n",
    "             \n",
    "hap_map.close()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 2",
   "language": "python",
   "name": "python2"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 2
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython2",
   "version": "2.7.11"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 0
}
